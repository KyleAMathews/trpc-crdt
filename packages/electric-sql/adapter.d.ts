import { AnyRouter } from '@trpc/server';

interface OnErrorParams {
    error: Error;
    path: string;
    type: string;
    ctx: any;
    input: any;
}
interface AdapterArgs {
    db: any;
    appRouter: AnyRouter;
    context: any;
    onError?: (params: OnErrorParams) => void;
}
declare function adapter({ appRouter, context, onError }: AdapterArgs): Promise<void>;

export { adapter };
