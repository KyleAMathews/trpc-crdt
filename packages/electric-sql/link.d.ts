import { AnyRouter } from '@trpc/server';
import { Doc } from 'yjs';
import { TRPCLink } from '@trpc/client';

declare const link: <TRouter extends AnyRouter>({ doc, }: {
    doc: Doc;
}) => TRPCLink<TRouter>;

export { link };
