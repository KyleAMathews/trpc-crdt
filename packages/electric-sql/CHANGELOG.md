# trpc-yjs

## 0.0.3

### Patch Changes

- Update electric-sql to user-controlled response writing pattern

## 0.0.2

### Patch Changes

- Add prepublish scripts

## 0.0.6

### Patch Changes

- Fix build

## 0.0.5

### Patch Changes

- Fix type of uuidv4 function and actually build

## 0.0.4

### Patch Changes

- The uuid package pulls in a bunch of extra stuff. crypto.getRandomValues is good enough

## 0.0.3

### Patch Changes

- Don't have yjs/trpc as dependencies â€” that's responsibility of implementor

## 0.0.2

### Patch Changes

- 14120dd: Initial setup of repo and release of trpc-yjs
- c72585d: Fixes and API cleanups
