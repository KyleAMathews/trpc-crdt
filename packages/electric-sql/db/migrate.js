const { DATABASE_URL, PUBLIC_DATABASE_URL } = require(`./util.js`)
const { spawnSync } = require(`child_process`)
const process = require(`process`)

console.info(`Connecting to postgres at ${PUBLIC_DATABASE_URL}`)

const args = [
  `exec`,
  // `-s`,
  `pg-migrations`,
  `apply`,
  `--database`,
  DATABASE_URL,
  `--directory`,
  `./db/migrations`,
]
console.log({ args })
const res = spawnSync(`pnpm`, args, { cwd: __dirname })

let newMigrationsApplied = true

console.log(res)
console.log(res.output.toString())

// proc.on(`exit`, (code) => {
// console.log(code)
// if (code === 0) {
// if (newMigrationsApplied) {
// console.log(`⚡️ Database migrated.`)
// } else {
// console.log(`⚡ Database already up to date.`)
// }
// }
// })
