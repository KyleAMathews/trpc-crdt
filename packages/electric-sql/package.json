{
  "name": "trpc-electric-sql",
  "description": "Call server functions natively in your ElectricSQL app with tRPC",
  "version": "0.1.1",
  "author": "Kyle Mathews <mathews.kyle@gmail.com>",
  "devDependencies": {
    "@databases/pg": "^5.5.0",
    "@databases/pg-migrations": "^5.0.2",
    "@electric-sql/prisma-generator": "1.1.0",
    "@prisma/client": "4.8.1",
    "@trpc/client": "^10.43.1",
    "@trpc/server": "^10.43.1",
    "@types/node": "^20.8.10",
    "@types/react": "^18.2.34",
    "@types/react-dom": "^18.2.14",
    "better-sqlite3": "^8.7.0",
    "electric-sql": "^0.7.0",
    "esbuild": "^0.16.17",
    "esbuild-plugin-inline-image": "^0.0.9",
    "fs-extra": "^10.1.0",
    "pg": "^8.11.3",
    "prisma": "4.8.1",
    "shelljs": "^0.8.5",
    "typescript": "^4.9.5",
    "vitest": "^0.34.6"
  },
  "exports": {
    "./adapter": "./dist/adapter.js",
    "./link": "./dist/link.js"
  },
  "files": [
    "dist",
    "*.d.ts"
  ],
  "keywords": [],
  "license": "MIT",
  "peerDependencies": {
    "@trpc/client": "*",
    "@trpc/server": "*",
    "electric-sql": ">0.7.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "backend:start": "docker compose --env-file backend/compose/.envrc -f backend/compose/docker-compose.yaml up",
    "backend:down": "docker compose -f backend/compose/docker-compose.yaml down --volumes",
    "db:migrate": "node ./db/migrate.js",
    "db:psql": "node ./db/connect.js",
    "client:generate": "yarn electric:check && npx electric-sql generate --service http://localhost:5133",
    "electric:check": "node ./backend/check-electric-is-running.js",
    "build": "npm run clean && tsup && npm run post-build",
    "prepublishOnly": "npm run build",
    "check": "tsc",
    "clean": "shx rm -rf dist *.d.ts",
    "post-build": "shx mv dist/*.d.ts* .",
    "test": "vitest"
  },
  "tsup": {
    "entry": [
      "src/adapter.ts",
      "src/link.ts"
    ],
    "format": [
      "esm",
      "cjs"
    ],
    "dts": true
  }
}
