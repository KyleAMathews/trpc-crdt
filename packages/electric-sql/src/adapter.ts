import { AnyRouter, callProcedure, getTRPCErrorFromUnknown } from "@trpc/server"
import { getErrorShape } from "@trpc/server/shared"

interface OnErrorParams {
  error: Error
  path: string
  type: string
  ctx: any
  input: any
}

// Define the expected types for the arguments
interface AdapterArgs {
  db: any
  appRouter: AnyRouter
  // eslint-disable-next-line
  context: any
  onError?: (params: OnErrorParams) => void
}

export async function adapter({
  electric,
  appRouter,
  context,
  onError,
}: AdapterArgs) {
  const { db } = electric
  // Handle new tRPC calls.
  async function handleCall(callObj) {
    const transactionFns: any[] = []
    const transact = (fn: () => void) => {
      transactionFns.push(fn)
    }
    const response = await callProcedure({
      procedures: appRouter._def.procedures,
      path: callObj.path,
      rawInput: JSON.parse(callObj.input),
      type: callObj.type,
      ctx: { ...context, electric, transact },
    })

    const transactionPromises = transactionFns.map((fn) => fn())

    // This won't happen in the same transaction but because they run
    // simultaneously, they should be synced back to postgres at the same time (I think).
    await Promise.all([
      ...transactionPromises,
      db.trpc_calls.update({
        data: {
          done: 1,
          response: JSON.stringify(response),
        },
        where: {
          id: callObj.id,
        },
      }),
    ])
  }

  const live = db.trpc_calls.liveMany({ where: { done: 0 } })

  const initialRes = await live()
  initialRes.result.forEach((callObj) => handleCall(callObj))

  electric.notifier.subscribeToDataChanges(async (event) => {
    const res = await live()
    if (res.result.length > 0) {
      res.result.forEach((callObj) => handleCall(callObj))
    }
  })

  // const requests = doc.getArray(`trpc-calls`)
  // requests.observe(async (event: Y.YArrayEvent<any>) => {
  // const { insert } = (event?.changes?.delta as any[]).find((item) => {
  // return `insert` in item
  // }) || { insert: [] }
  // insert.forEach(async (state: any) => {
  // // Backwards compatability for older calls that were plain objects.
  // if (!state.get) {
  // return
  // }
  // if (state.get(`done`) !== true) {
  // const transactionFns: any[] = []
  // const transact = (fn: () => void) => {
  // transactionFns.push(fn)
  // }
  // try {
  // const response = await callProcedure({
  // procedures: appRouter._def.procedures,
  // path: state.get(`path`),
  // rawInput: state.get(`input`),
  // type: state.get(`type`),
  // ctx: { ...context, transact },
  // })
  // doc.transact(() => {
  // transactionFns.forEach((fn) => {
  // fn()
  // })
  // state.set(`response`, response)
  // state.set(`done`, true)
  // })
  // } catch (cause) {
  // const error = getTRPCErrorFromUnknown(cause)
  // const errorShape = getErrorShape({
  // config: appRouter._def._config,
  // error,
  // type: state.get(`type`),
  // path: state.get(`path`),
  // input: state.get(`input`),
  // ctx: context,
  // })

  // onError?.({
  // error,
  // path: state.get(`path`),
  // type: state.get(`type`),
  // ctx: context,
  // input: state.get(`input`),
  // })

  // doc.transact(() => {
  // state.set(`done`, true)
  // state.set(`error`, true)
  // state.set(`response`, { error: errorShape })
  // })
  // }
  // }
  // })
  // })
}
