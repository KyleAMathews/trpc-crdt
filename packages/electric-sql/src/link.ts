import type { AnyRouter } from "@trpc/server"
import { TRPCClientError, TRPCLink } from "@trpc/client"
import { observable } from "@trpc/server/observable"
import { genUUID } from "electric-sql/util"

export const link = <TRouter extends AnyRouter>({
  electric,
  clientId,
}): TRPCLink<TRouter> => {
  const { db } = electric
  const live = db.trpc_calls.liveMany({
    where: { clientid: clientId, done: 1 },
  })

  const requestMap = new Map()
  async function observe() {
    const res = await live()
    if (res.result.length > 0) {
      res.result.forEach((callRes) => {
        if (requestMap.has(callRes.id)) {
          requestMap.get(callRes.id)(callRes)
          requestMap.delete(callRes.id)
        }
      })
    }
  }

  electric.notifier.subscribeToDataChanges(observe)

  return () =>
    ({ op }) =>
      observable(async (observer) => {
        const requestId = genUUID()

        requestMap.set(requestId, (callRes) => {
          const elapsedMs =
            new Date().getTime() -
            new Date((callRes.createdat as number) || 0).getTime()

          observer.next({
            result: {
              type: `data`,
              data: JSON.parse(callRes.response),
            },
          })
          db.trpc_calls.update({
            data: {
              elapsedms: elapsedMs,
            },
            where: {
              id: requestId,
            },
          })
        })

        // TODO stop listening
        // handle errors

        // The observe function to listen to the response
        // from the server.
        // function observe(event: YMapEvent<any>) {
        // const state = event.target
        // if (state.get(`done`) && state.get(`id`) === requestId) {
        // requestMap.unobserve(observe)
        // requestMap.set(
        // `elapsedMs`,
        // new Date().getTime() -
        // new Date((requestMap.get(`createdAt`) as number) || 0).getTime()
        // )
        // if (state.get(`error`)) {
        // observer.error(TRPCClientError.from(state.get(`response`)))
        // } else {
        // observer.next({
        // result: {
        // type: `data`,
        // data: state.get(`response`),
        // },
        // })
        // }
        // observer.complete()
        // }
        // }

        // Create the request map on the trpc-calls Y.js Array.
        // This will get replicated to the server.
        const { path, input, type } = op

        await db.trpc_calls.create({
          data: {
            id: requestId,
            path,
            input: JSON.stringify(input),
            type,
            done: 0,
            error: 0,
            createdat: new Date().toJSON(),
            clientid: clientId,
          },
        })
      })
}
