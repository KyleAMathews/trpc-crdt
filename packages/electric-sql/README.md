# trpc-electric-sql

[tRPC](https://trpc.io/) integration for [ElectricSQL](https://electric-sql.com/): ElectricSQL-native RPC calls

For when you're building with ElectricSQL but you want to run code on a server.

See the [monorepo README for a longer writeup on why this library exists](https://github.com/KyleAMathews/trpc-crdt).

## Install

`npm install trpc-electric-sql`

## Example app

https://github.com/KyleAMathews/trpc-crdt/tree/main/examples/electric-sql

## Usage

This integration with ElectricSQL serializes tRPC calls via an electrified `trpc_calls` table.

Add this table definition to your Postgres instance:

```sql
CREATE TABLE IF NOT EXISTS trpc_calls (
    id UUID PRIMARY KEY NOT NULL,
    createdAt TIMESTAMPTZ NOT NULL,
    elapsedMs INTEGER,
    path TEXT NOT NULL,
    input TEXT,
    type TEXT NOT NULL,
    state TEXT NOT NULL,
    clientId TEXT NOT NULL,
    response TEXT
);

ALTER TABLE trpc_calls ENABLE ELECTRIC;
```

Once [DDLX support lands](https://electric-sql.com/docs/api/ddlx), we'll be
able to restrict updates to trpc_calls rows to the server and the original
creator.

You can then setup your browser/server code similar to the following.

#### Browser

```ts
import { createTRPCProxyClient } from "@trpc/client"
import { link, createElectricRef } from "trpc-electric-sql/link"
import { Electric, schema } from "../src/generated/client"
import { ElectricDatabase, electrify } from "electric-sql/wa-sqlite"
import { genUUID, uniqueTabId } from "electric-sql/util"
import { authToken } from "../auth"

const electricRef = createElectricRef<Electric>()
const trpc = createTRPCProxyClient<AppRouter>({
  links: [
    link({
      electricRef,
      clientId: genUUID(),
    }),
  ],
})

const init = async () => {
  const token = authToken()
  const config = {
    auth: {
      token: token,
    },
  }

  const { tabId } = uniqueTabId()
  const tabScopedDbName = `electric-${tabId}.db`

  const conn = await ElectricDatabase.init(tabScopedDbName, ``)
  const electric = await electrify(conn, schema, config)

  const [shape, usersShape] = await Promise.all([
    electric.db.trpc_calls.sync(),
    electric.db.users.sync(),
  ])
  await Promise.all([shape.synced, usersShape.synced])

  electricRef.value = electric
  setElectric(electric)

  // Tell server to create a new user.
  await trpc.userCreate.mutation({id: `1`, name: `Kyle Mathews`})

  // The new user, written by the server, is now available in the local sqlite db:
  const user = await electric.db.users.findUnique({where: { id: `1` }})
}

init()

```

#### Server

```typescript
import { adapter } from "trpc-electric-sql/adapter"
import { initTRPC, TRPCError } from "@trpc/server"
import { z } from "zod"

const t = initTRPC.create()
const router = t.router
const publicProcedure = t.procedure

const appRouter = router({
  userCreate: publicProcedure
    .input(z.object({ name: z.string(), id: z.string() }))
    .mutation(async (opts) => {
      const {
        input,
        ctx: {
          transact,
          electric: { db },
        },
      } = opts
      const user = { ..input }

      // Writes in the transact function gets applied at same time the trpc call
      // is finished.
      transact(() => {
        db.users.create({
          data: user,
        })
      })
    })
})

async function setupTRPC() {
  const config = {
    auth: {
      token: authToken(),
    },
    url: process.env.ELECTRIC_URL,
  }

  // Create the better-sqlite3 database connection.
  const conn = new Database(`local-data.db`)
  conn.pragma(`journal_mode = WAL`)

  // Instantiate your electric client.
  const electric = await electrify(conn, schema, config)
  const { db } = electric
  const [shape, usersShape] = await Promise.all([
    electric.db.trpc_calls.sync(),
    electric.db.users.sync(),
  ])
  await Promise.all([shape.synced, usersShape.synced])

  adapter({
    context: { electric },
    appRouter,
  })
}

setupTRPC()
```
