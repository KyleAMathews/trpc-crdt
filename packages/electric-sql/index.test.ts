import { describe, it, expect, afterAll } from "vitest"
import { initTRPC, TRPCError } from "@trpc/server"
import { z } from "zod"
import shelljs from "shelljs"
import { genUUID } from "electric-sql/util"
import { adapter } from "./src/adapter"
import { link } from "./src/link"

import Database from "better-sqlite3"
import { electrify } from "electric-sql/node"
import { authToken } from "./auth"
import { schema } from "./src/generated/client"
import { DATABASE_URL } from "./db/util"
import { Client } from "pg"

const { DEBUG_MODE, ELECTRIC_URL } = {
  DEBUG_MODE: false, //true, //process.env.DEBUG_MODE === "true",
  ELECTRIC_URL: process.env.ELECTRIC_URL ?? `ws://localhost:5133`,
}

import { createTRPCProxyClient, loggerLink, httpBatchLink } from "@trpc/client"
/**
 * Initialization of tRPC backend
 * Should be done only once per backend!
 */
const t = initTRPC.create()
/**
 * Export reusable router and procedure helpers
 * that can be used throughout the router
 */
const router = t.router
const publicProcedure = t.procedure

async function makeClientTable(dbName: string, isServer = true) {
  const config = {
    auth: {
      token: authToken(),
    },
    debug: DEBUG_MODE,
    url: ELECTRIC_URL,
  }

  // Create the better-sqlite3 database connection. The first
  // argument is your database name. Changing this will
  // create/use a new local database file.
  const conn = new Database(
    `sqliteDbs/${dbName}-${isServer ? `server` : `client`}.db`
  )

  // Follow the library recommendation to enable WAL mode to
  // increase performance. As per:
  // https://github.com/WiseLibs/better-sqlite3/blob/master/docs/performance.md
  conn.pragma(`journal_mode = WAL`)

  // Instantiate your electric client.
  const electric = await electrify(conn, schema, config)
  const { db } = electric
  const shape = await db.trpc_calls.sync()
  const usersShape = await db.users.sync()
  await shape.synced
  await usersShape.synced

  console.log(
    `created client db sqliteDbs/${dbName}-${isServer ? `server` : `client`}.db`
  )

  return electric
}

async function initClient(dbName) {
  console.log(`initClient`, dbName)
  // console.trace()
  //
  const [serverElectric, clientElectric] = await Promise.all([
    makeClientTable(dbName, true),
    makeClientTable(dbName, false),
  ])

  const pgClient = new Client({ connectionString: DATABASE_URL })
  await pgClient.connect()

  // Start adapter
  const appRouter = router({
    userCreate: publicProcedure
      .input(
        z.object({ name: z.string(), optionalDelay: z.number().optional() })
      )
      .mutation(async (opts) => {
        const {
          input,
          ctx: {
            transact,
            electric: { db },
          },
        } = opts
        const users = await db.users.findMany()
        const user = { id: String(users.length + 1), ...input }

        if (input.optionalDelay) {
          await new Promise((resolve) =>
            setTimeout(resolve, input.optionalDelay)
          )
        }

        if (input.name === `BAD_NAME`) {
          throw new TRPCError({
            code: `CONFLICT`,
            message: `This name isn't one I like to allow`,
          })
        }

        // Run in transaction along with setting response on the request
        // object.
        transact(() =>
          db.users.create({
            data: {
              id: user.id.toString(),
              name: user.name,
            },
          })
        )

        return user
      }),
    userUpdateName: publicProcedure
      .input(z.object({ id: z.string(), name: z.string() }))
      .mutation(async (opts) => {
        const {
          input,
          ctx: {
            transact,
            electric: { db },
          },
        } = opts
        // Run in transaction along with setting response on the request
        // object.
        transact(() => {
          return db.users.update({
            data: {
              name: input.name,
            },
            where: {
              id: input.id,
            },
          })
        })

        return `ok`
      }),
  })

  let trpc
  if (dbName !== `cleanup`) {
    type AppRouter = typeof appRouter
    adapter({ electric: serverElectric, appRouter })

    // Create client.
    trpc = createTRPCProxyClient<AppRouter>({
      links: [
        link({
          electric: clientElectric,
          clientId: genUUID(),
        }),
      ],
    })
  }

  // return { doc: clientDoc, trpc }
  return { clientElectric, serverElectric, pgClient, trpc }
}

afterAll(async () => {
  const { pgClient } = await initClient(`cleanup`)
  await Promise.all([
    pgClient.query(`DELETE from trpc_calls`),
    pgClient.query(`DELETE from users`),
    shelljs.rm(`./sqliteDbs/*`),
  ])
})

describe(`electric-sql`, () => {
  describe(`basic db connectivity`, () => {
    it(`can connect to server sqlite and create a trpc call`, async () => {
      const {
        clientElectric: { db },
      } = await initClient(`basic-connectivity`)
      expect(true).toBeTruthy()
      const id = genUUID()
      await db.users.create({
        data: {
          id,
          name: `foo`,
        },
      })
      const res = await db.users.findMany()
      expect(res).toHaveLength(1)
      expect(res[0].id).toEqual(id)
    })
  })
  describe(`basic calls`, async () => {
    const {
      trpc,
      clientElectric: { db },
    } = await initClient(`basic-calls`)
    it(`create a user`, async () => {
      const newUser = await trpc.userCreate.mutate({ name: `foo` })
      expect(newUser.name).toEqual(`foo`)
      const user = await db.users.findUnique({ where: { id: `2` } })
      expect(user).toMatchSnapshot()
      expect(user.name).toEqual(`foo`)
    })
    it(`updateName`, async () => {
      await trpc.userUpdateName.mutate({ id: `2`, name: `foo2` })
      const user = await db.users.findUnique({ where: { id: `2` } })
      expect(user.name).toEqual(`foo2`)
    })
  })
  // describe(`batched calls`, async () => {
  // const { doc, trpc } = await initClient(`batched-calls`)
  // it(`handles batched calls`, async () => {
  // const promise1 = trpc.userCreate.mutate({ name: `foo1` })
  // const promise2 = trpc.userCreate.mutate({ name: `foo2` })

  // await Promise.all([promise1, promise2])

  // const promise3 = trpc.userCreate.mutate({ name: `foo3` })
  // const promise4 = trpc.userCreate.mutate({ name: `foo4` })

  // await Promise.all([promise3, promise4])

  // await trpc.userCreate.mutate({ name: `foo5` })

  // const users = await db.users.findMany()
  // // const users = doc.getArray(`users`).toJSON()

  // expect(users).toHaveLength(7)
  // })
  // })
  // describe(`out-of-order calls`, async () => {
  // const { trpc, doc } = initClient(`out-of-order`)
  // it(`handles out-of-order calls`, async () => {
  // const user1Promise = trpc.userCreate.mutate({
  // name: `foo1`,
  // optionalDelay: 10,
  // })
  // const user2Promise = trpc.userCreate.mutate({ name: `foo2` })
  // const [user1, user2] = await Promise.all([user1Promise, user2Promise])
  // expect(user1.name).toEqual(`foo1`)
  // expect(user2.name).toEqual(`foo2`)
  // })
  // })
  // describe(`handle errors`, () => {
  // const { trpc } = initClient(`handle-errors`)
  // it(`input errors`, async () => {
  // await expect(() =>
  // trpc.userCreate.mutate({ name: 1 })
  // ).rejects.toThrowError(`invalid_type`)
  // })
  // it(`router thrown errors`, async () => {
  // await expect(() =>
  // trpc.userCreate.mutate({ name: `BAD_NAME` })
  // ).rejects.toThrowError(`This name isn't one I like to allow`)
  // })
  // })
})
