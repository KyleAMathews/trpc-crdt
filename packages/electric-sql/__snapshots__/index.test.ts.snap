// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`electric-sql > basic calls > create a user 1`] = `
{
  "id": "2",
  "name": "foo",
}
`;
